### ISSUE TITLE
Bug in calculation of Toomre's Q

### ISSUE THREAD

**Issue by ttepperg** [pynbody/pynbody#562]
In analysis/profile.py, the stability parameter Q is currently calculated as:

    return (self['vr_disp'] * self['kappa'] / (3.36 * self['density'] * units.G)).in_units("")

However, I'm certain that 'vrxy' should be used instead of 'vr', i.e.

    return (self['vrxy_disp'] * self['kappa'] / (3.36 * self['density'] * units.G)).in_units("")

The quantity 'vr' is the (spherical) radial velocity whereas 'vrxy' is the (cylindrical)
mid-plane radial velocity. The latter is the one actually entering the definition of
Toomre's Q.

**Comment by apontzen** [comment]
Thanks for pointing this out, I agree

**Comment by ttepperg** [comment]
Hi Andrew,

On a second look, I realised that the problem with the function mentioned above (Toomre’s Q) is more serious than I first thought and its solution far more complex.

The problem is that in the definition of Q, the (cylindrical) radial velocity dispersion and the surface density refer to the disc component, while the angular frequency, omega, and thus the epicyclic frequency, kappa, depend on the *total* potential of the system, say, host halo, bulge, and disc.

The solution I first suggested is fine for a self-gravitating (i.e. fully isolated) disc. But this case is not generally of interest. The more interesting case of a disc embedded within a DM halo and perhaps with a bulge will require a careful calculation of the relevant quantities taking each component into account. I have followed the latter approach but my function may not be useful in general, as I have tailored it to deal with my very particular Ramses simulations. I’ll more than happy to share it but it would definitely need to be modified before it can be embedded within pynbody.

For the time being, I'd recommend fully removing the calculation of Toomre’s Q from pynbody's suite of profile functions, or at least issue a warning to the user of the potential pitfall.

Cheers,

Thor.

**Comment by ttepperg** [comment]
Hi Andrew,

I just posted a new comment on this.

Cheers,

Thor.

**Comment by apontzen** [comment]
That makes sense but would be interested to get @rokroskar’s take on this before actually removing.

@ttepperg it is always much appreciated when people can contribute enhancements back to the project. Please do consider taking the time to make your code generally applicable and then open a pull request!

**Comment by ttepperg** [comment]
HI Andrew,

Nothing would make me happier than contributing to Pynbody.

The issue with my scripts right now is that they are very specific to my simulations. In these I know exactly what particles make up a disc, a halo, a bulge. So I do not make use of any decomposition algorithms.

If you can tell me how to generally distinguish in any galaxy simulation between disc, halo, etc. and provide an example so I can test it, I’ll be thrilled to reshape my Toomre_Q routine into something that can be integrated within Pynbody.

Cheers,

Thor.

**Comment by rokroskar** [comment]
Thanks for pointing this out @ttepperg - one thing I can't remember (and have no quick way of discerning right now) is whether the gravity calculation (which is used in the determination of kappa) takes into account all of the particles or just the sub-component used in the profile. If the former then this is safe, if the latter than I agree it should probably be removed until a better solution can be found.

**Comment by ttepperg** [comment]
Oh, I see, good point. I'm afraid I do not know that either. Andrew may have to wage in again and say.
But even if the potential is correctly taken into account, vr_disp definitely needs to be replaced by vrxy_disp.
That the disk is already aligned with (i.e. in) the xy plane, vr_disp and vrxy_disp may differer significantly from one another
if vz_disp is non-negligible.

**Comment by rokroskar** [comment]
I agree that in-plane and radial velocity dispersions may be quite different depending on the disk structure, but I thought that that term in the calculation of Toomre Q was supposed to be the thermal pressure? I don't have references at the top of my head anymore, but I seem to recall most literature using `sigma_r` and not `sigma_xy`.

**Comment by ttepperg** [comment]
Hi,

Thanks for adding to the discussion!

All of the literature I know (see Binney&Tremaine, always trustworthy) uses sigma_R, where ‘R' is generally used to denote the cylindrical (as opposed to spherical) radius, i.e. the radius on the xy-plane.  It makes sense because the stability criterion is all about in-plane motions, and how these work against the gravitational field in the plane. The latter is reflected in the fact that the surface density, and not the volume density, enters into the calculation of Q. The vertical motions do not play a role, as far as I know. In a way, the z component of the disc has been marginalised over.

In the case of gaseous discs, sigma_R is in fact replaced with the sound speed c_s (directly related to the thermal pressure), but in this case the situation is physically different, because the pressure is isotropic (neglecting magnetic fields, etc.) This is also reflected in the slightly different numerical factor.

I’m almost 100% sure that the velocity dispersion entering Toomre’s stability criterion is sigma_R (not sigma_r). I could be mistaken, however, so I guess it is up to the developers to decide whether this is a bug or not.

Cheers,

Thor.

**Comment by apontzen** [comment]
I agree it should be sigma_R. However, just to clarify - sigma_r is not the 3D dispersion. It is the dispersion along the radial direction. So, if the majority of the mass is in a disk, it will be very similar to sigma_R. The advantage of the current implementation is that it doesn’t care about which orientation the disk is in - it’ll always give a roughly right answer (provided the mass is concentrated as above). Overall I favour switching to sigma_R and issuing a warning that the disk must be oriented correctly (as other routines already do). The question regarding gravity is more subtle, I’ll take a look through the code and see if I can figure it out.

**Comment by ttepperg** [comment]
Sounds good. Thanks a lot, Andrew.

**Comment by apontzen** [comment]
Having looked at it, I think it was only using the gravity contributed by the profiled component. I've now updated this in branch `toomre-q` (PR #566). Would appreciate you both checking it over @rokroskar @ttepperg as this is not something I have ever used myself, and therefore I don't have a great test scenario.

**Comment by ttepperg** [comment]
Andrew,

Thanks so much for taking the time to do this,
I’ll test the code and let you know how it goes.

Thor.

**Comment by apontzen** [comment]
@ttepperg quick ping to see if you had a chance to check this? It would be great to close this issue.

**Comment by ttepperg** [comment]
Hi Andrew,

Sorry, I have not; I’ve been away for a while and now things
are crazy around here. I’ll do that ASAP

Thor.


PS: Just about to submit a paper where we, as usual, acknowledge
Pynbody ;)

**Comment by ttepperg** [comment]
Hi Andrew,

Ok, I checked the new calculation of Toomre’s Q in branch toomre-q.

I ran several test, and there seems to be an *exact* factor of 10 mismatch,
i.e. I can recover the result I calculate by hand (or using other codes) with
Pynbody’s ‘Q’ profile if I scale the latter by 1/10 (plot attached).
I am sure that my result is correct because I have tested and re-retested,
and it matches independent calculation.

Unfortunately, I could not isolate the problem further, for if I compare each of
the individual quantities that go into the calculation of Q: surface density,
sigma_R, kappa, even v_circ_total and omega, they ALL match my own results.

Since Q is dimensionless, I am really puzzled...There must be a strange conversion
factor sneaking in sowewhere, but I have no idea where. I just noticed that kappa
is output in Gyr^-1 and perhaps it is implicitly assumed that it should be Myr^-1 ?
However, that yields a factor of 1000 difference and there is no cubic root anywhere…


Thor


_______________________________________________
Dr. Thorsten Tepper García | Astrophysicist
Centre for Integrated Sustainability Analysis (ISA)
Sydney Institute for Astronomy (SIfA)
The University of Sydney
www.thorsten.mx

**Comment by apontzen** [comment]
Thanks for checking. This sounds very strange. 

I don't think github has correctly attached your image unfortunately.

Ideally could you post a comparison of results for kappa for the pynbody testdata galaxy? This would allow for a reproducible test and so help getting to the bottom of it.

**Comment by ttepperg** [comment]
Hi Andrew,

> Ideally could you post a comparison of results for kappa for the pynbody testdata galaxy? This would allow for a reproducible test and so help getting to the bottom of it.
> 
Not sure which testdata galaxy you are referring to.

By comparison you mean using both the ‘old’ master version and branch toomre-q
to calculate kappa?

**Comment by apontzen** [comment]
I mean g15784, as referenced here http://pynbody.github.io/pynbody/tutorials/snapshot_manipulation.html,  http://star.ucl.ac.uk/~app/testdata.tar.gz. Thanks!
